[Offload][NFC] use unique ptrs for platforms

pbalcer opened the pull request:
Currently, devices store a raw pointer to back to their owning Platform. Platforms are stored directly inside of a vector. Modifying this vector risks invalidating all the platform pointers stored in devices.

This patch allocates platforms individually, and changes devices to store a reference to its platform instead of a pointer. This is safe, because platforms are guaranteed to outlive the devices they contain.

jhuber6 reviewed (approved):
This looks good to me assuming it passes the unit tests, but gives @rossburton a chance to look through it.

pbalcer committed:
[Offload][NFC] use unique ptrs for platforms

Currently, devices store a raw pointer to back to their owning
Platform. Platforms are stored directly inside of a vector. Modifying
this vector risks invalidating all the platform pointers stored in
devices.

This patch allocates platforms individually, and changes devices to
store a reference to its platform instead of a pointer. This is safe,
because devices are guaranteed to outlive the platform they are from.
